## Importing modules and loading data

In [1]:
import pandas as pd
import numpy as np

In [2]:
logins =pd.read_json('logins.json')

In [60]:
logins.head()

login_time
0 1970-01-01 20:13:18
1 1970-01-01 20:16:10
2 1970-01-01 20:16:37
3 1970-01-01 20:16:36
4 1970-01-01 20:26:21

In [19]:
type(logins)

pandas.core.frame.DataFrame

This data consists of times of login only. We can use this to analyse the trends in the logins across given timeframe.

In [3]:
ultimate =pd.read_json('ultimate_data_challenge.json')

In [79]:
ultimate.head(10)

city  trips_in_first_30_days signup_date  avg_rating_of_driver  \
0  King's Landing                       4  2014-01-25                   4.7   
1         Astapor                       0  2014-01-29                   5.0   
2         Astapor                       3  2014-01-06                   4.3   
3  King's Landing                       9  2014-01-10                   4.6   
4      Winterfell                      14  2014-01-27                   4.4   
5      Winterfell                       2  2014-01-09                   3.5   
6         Astapor                       1  2014-01-24                   NaN   
7      Winterfell                       2  2014-01-28                   5.0   
8      Winterfell                       2  2014-01-21                   4.5   
9      Winterfell                       1  2014-01-03                   NaN   

   avg_surge last_trip_date    phone  surge_pct  ultimate_black_user  \
0       1.10     2014-06-17   iPhone       15.4                 True   
1       1.00     2014-05-05  Android        0.0                False   
2       1.00     2014-01-07   iPhone        0.0                False   
3       1.14     2014-06-29   iPhone       20.0                 True   
4       1.19     2014-03-15  Android       11.8                False   
5       1.00     2014-06-06   iPhone        0.0                 True   
6       1.00     2014-01-25  Android        0.0                False   
7       1.00     2014-01-29   iPhone        0.0                False   
8       1.00     2014-02-01  Android        0.0                False   
9       1.00     2014-01-05  Android        0.0                False   

   weekday_pct  avg_dist  avg_rating_by_driver  
0         46.2      3.67                   5.0  
1         50.0      8.26                   5.0  
2        100.0      0.77                   5.0  
3         80.0      2.36                   4.9  
4         82.4      3.13                   4.9  
5        100.0     10.56                   5.0  
6        100.0      3.95                   4.0  
7        100.0      2.04                   5.0  
8        100.0      4.36                   5.0  
9          0.0      2.37                   5.0

In [21]:
ultimate.shape

(50000, 12)

In [22]:
ultimate.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 12 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   city                    50000 non-null  object 
 1   trips_in_first_30_days  50000 non-null  int64  
 2   signup_date             50000 non-null  object 
 3   avg_rating_of_driver    41878 non-null  float64
 4   avg_surge               50000 non-null  float64
 5   last_trip_date          50000 non-null  object 
 6   phone                   49604 non-null  object 
 7   surge_pct               50000 non-null  float64
 8   ultimate_black_user     50000 non-null  bool   
 9   weekday_pct             50000 non-null  float64
 10  avg_dist                50000 non-null  float64
 11  avg_rating_by_driver    49799 non-null  float64
dtypes: bool(1), float64(6), int64(1), object(4)
memory usage: 4.2+ MB


In [23]:
ultimate.describe()

trips_in_first_30_days  avg_rating_of_driver     avg_surge  \
count            50000.000000          41878.000000  50000.000000   
mean                 2.278200              4.601559      1.074764   
std                  3.792684              0.617338      0.222336   
min                  0.000000              1.000000      1.000000   
25%                  0.000000              4.300000      1.000000   
50%                  1.000000              4.900000      1.000000   
75%                  3.000000              5.000000      1.050000   
max                125.000000              5.000000      8.000000   

          surge_pct   weekday_pct      avg_dist  avg_rating_by_driver  
count  50000.000000  50000.000000  50000.000000          49799.000000  
mean       8.849536     60.926084      5.796827              4.778158  
std       19.958811     37.081503      5.707357              0.446652  
min        0.000000      0.000000      0.000000              1.000000  
25%        0.000000     33.300000      2.420000              4.700000  
50%        0.000000     66.700000      3.880000              5.000000  
75%        8.600000    100.000000      6.940000              5.000000  
max      100.000000    100.000000    160.960000              5.000000

This data is at user level. we can identify each user from the index of the row.

## Data Cleansing

### Replacing missing values

In [66]:
logins['login_time'].isnull().sum()

0

In [24]:
logins.describe()

C:\Users\B.K.M.Bhat\Anaconda\lib\site-packages\ipykernel_launcher.py:1: FutureWarning: Treating datetime data as categorical rather than numeric in `.describe` is deprecated and will be removed in a future version of pandas. Specify `datetime_is_numeric=True` to silence this warning and adopt the future behavior now.
  """Entry point for launching an IPython kernel.


login_time
count                 93142
unique                92265
top     1970-02-12 11:16:53
freq                      3
first   1970-01-01 20:12:16
last    1970-04-13 18:57:38

In [67]:
ultimate.isnull().sum()

city                         0
trips_in_first_30_days       0
signup_date                  0
avg_rating_of_driver      8122
avg_surge                    0
last_trip_date               0
phone                      396
surge_pct                    0
ultimate_black_user          0
weekday_pct                  0
avg_dist                     0
avg_rating_by_driver       201
dtype: int64

Average rating of driver is seen to be missing for 8122 records. Average rating by driver is missing for 201 records. 
These values can be imputed. Let's find out unique values in these cases and go ahead with the imputation.

In [4]:
ultimate['avg_rating_by_driver'][ultimate['avg_rating_by_driver'].isnull()]

90      NaN
274     NaN
277     NaN
536     NaN
857     NaN
         ..
49249   NaN
49376   NaN
49504   NaN
49599   NaN
49626   NaN
Name: avg_rating_by_driver, Length: 201, dtype: float64

In [5]:
ultimate['avg_rating_by_driver'].value_counts()

5.0    28508
4.8     4537
4.7     3330
4.9     3094
4.5     2424
4.6     2078
4.0     1914
4.3     1018
4.4      860
3.0      602
4.2      342
3.5      199
3.7      195
1.0      181
2.0      126
4.1      125
3.8      111
3.3       47
3.9       41
2.5       31
3.6       19
3.4        5
1.5        4
2.8        3
3.2        2
2.7        2
2.3        1
Name: avg_rating_by_driver, dtype: int64

The rating falls between 1 and 5 for both the cases. Let's impute the missing values with the means.

In [6]:
ultimate['avg_rating_by_driver'].replace(np.nan, ultimate['avg_rating_by_driver'].mean(), inplace =True)

In [7]:
ultimate['avg_rating_by_driver'].isnull().sum()

0

In [8]:
ultimate['avg_rating_of_driver'].replace(np.nan, ultimate['avg_rating_of_driver'].mean(), inplace =True)

In [9]:
ultimate['avg_rating_of_driver'].isnull().sum()

0

Let's explore phone OS and replace missing phone OS with frequent OS

In [10]:
ultimate['phone'].unique()

array(['iPhone', 'Android', None], dtype=object)

In [11]:
ultimate['phone'].value_counts()

iPhone     34582
Android    15022
Name: phone, dtype: int64

Since iphones are more common, we can replace the missing OSs with iPhone

In [12]:
ultimate['phone'].replace(np.nan, 'iPhone', inplace =True)

In [13]:
ultimate['phone'].unique()

array(['iPhone', 'Android'], dtype=object)

### Exploring the remaining variables

In [14]:
ultimate['city'].unique()

array(["King's Landing", 'Astapor', 'Winterfell'], dtype=object)

There are 3 cities in which the services operate.

In [15]:
sorted(ultimate['trips_in_first_30_days'].unique())

[0,
 1,
 2,
 3,
 4,
 5,
 6,
 7,
 8,
 9,
 10,
 11,
 12,
 13,
 14,
 15,
 16,
 17,
 18,
 19,
 20,
 21,
 22,
 23,
 24,
 25,
 26,
 27,
 28,
 29,
 30,
 31,
 32,
 33,
 34,
 35,
 36,
 37,
 38,
 39,
 40,
 41,
 42,
 43,
 44,
 47,
 48,
 49,
 50,
 51,
 53,
 54,
 55,
 56,
 58,
 63,
 71,
 73,
 125]

Customers have had none to 125 trips in the past. On an average each customer has done 2 trips as obtained at the data description step.

In [16]:
ultimate['signup_date'].value_counts()

2014-01-18    2948
2014-01-25    2885
2014-01-11    2402
2014-01-24    2284
2014-01-17    2149
2014-01-31    2100
2014-01-19    2028
2014-01-10    2021
2014-01-06    1763
2014-01-01    1737
2014-01-26    1708
2014-01-23    1606
2014-01-07    1486
2014-01-04    1485
2014-01-30    1471
2014-01-09    1433
2014-01-16    1431
2014-01-22    1369
2014-01-05    1343
2014-01-12    1334
2014-01-20    1295
2014-01-28    1284
2014-01-08    1275
2014-01-27    1236
2014-01-21    1234
2014-01-03    1213
2014-01-29    1197
2014-01-14    1120
2014-01-15    1110
2014-01-13    1049
2014-01-02    1004
Name: signup_date, dtype: int64

In [17]:
ultimate['last_trip_date'].value_counts()

2014-06-29    2036
2014-06-28    1679
2014-06-30    1408
2014-06-27    1120
2014-06-22    1024
              ... 
2014-03-11      79
2014-02-19      79
2014-02-20      74
2014-03-10      69
2014-03-04      69
Name: last_trip_date, Length: 182, dtype: int64

In [18]:
ultimate['avg_surge'].value_counts()

1.00    34454
1.25     1100
1.13      956
1.02      809
1.08      798
        ...  
2.20        1
2.19        1
3.63        1
1.51        1
2.07        1
Name: avg_surge, Length: 115, dtype: int64

In [19]:
ultimate['surge_pct'].value_counts()

0.0      34409
100.0     1416
50.0      1367
33.3      1152
25.0       906
         ...  
27.9         1
34.7         1
62.5         1
29.3         1
26.4         1
Name: surge_pct, Length: 367, dtype: int64

In [20]:
ultimate['ultimate_black_user'].value_counts()         

False    31146
True     18854
Name: ultimate_black_user, dtype: int64

In [21]:
ultimate['weekday_pct'].value_counts() 

100.0    16659
0.0       9203
50.0      4057
66.7      2088
33.3      1619
         ...  
5.9          1
93.2         1
42.6         1
39.0         1
49.1         1
Name: weekday_pct, Length: 666, dtype: int64

In [22]:
ultimate['avg_dist'].value_counts() 

0.00     150
2.30     116
2.29     116
2.36     114
2.73     114
        ... 
28.85      1
21.97      1
33.14      1
39.31      1
30.64      1
Name: avg_dist, Length: 2908, dtype: int64

### Additional details to aide EDA

In [23]:
logins.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 93142 entries, 0 to 93141
Data columns (total 1 columns):
 #   Column      Non-Null Count  Dtype         
---  ------      --------------  -----         
 0   login_time  93142 non-null  datetime64[ns]
dtypes: datetime64[ns](1)
memory usage: 727.8 KB


In [24]:
logins['login_time'].value_counts()

1970-02-12 11:16:53    3
1970-02-24 23:09:57    3
1970-03-29 03:24:57    3
1970-03-14 17:45:52    3
1970-01-09 01:31:25    3
                      ..
1970-03-08 11:52:16    1
1970-01-08 09:30:07    1
1970-01-22 21:54:51    1
1970-01-08 23:27:50    1
1970-03-19 10:10:37    1
Name: login_time, Length: 92265, dtype: int64

The datstamps can help in identifying trends. Hence we can create a new column with datestamps only.

In [25]:
import datetime as dt
logins['login_date'] = logins['login_time'].dt.date

In [26]:
logins.head(10)

login_time  login_date
0 1970-01-01 20:13:18  1970-01-01
1 1970-01-01 20:16:10  1970-01-01
2 1970-01-01 20:16:37  1970-01-01
3 1970-01-01 20:16:36  1970-01-01
4 1970-01-01 20:26:21  1970-01-01
5 1970-01-01 20:21:41  1970-01-01
6 1970-01-01 20:12:16  1970-01-01
7 1970-01-01 20:35:47  1970-01-01
8 1970-01-01 20:35:38  1970-01-01
9 1970-01-01 20:47:52  1970-01-01

In [27]:
logins['login_time_extracted'] = logins['login_time'].dt.time

In [28]:
logins.head(10)

login_time  login_date login_time_extracted
0 1970-01-01 20:13:18  1970-01-01             20:13:18
1 1970-01-01 20:16:10  1970-01-01             20:16:10
2 1970-01-01 20:16:37  1970-01-01             20:16:37
3 1970-01-01 20:16:36  1970-01-01             20:16:36
4 1970-01-01 20:26:21  1970-01-01             20:26:21
5 1970-01-01 20:21:41  1970-01-01             20:21:41
6 1970-01-01 20:12:16  1970-01-01             20:12:16
7 1970-01-01 20:35:47  1970-01-01             20:35:47
8 1970-01-01 20:35:38  1970-01-01             20:35:38
9 1970-01-01 20:47:52  1970-01-01             20:47:52

In [29]:
logins.to_csv('login_details.csv', index=False)

In [32]:
ultimate.to_csv('ultimate_cleansed.csv', index=False)